yt_dlp/output/logging.py

```diff
@@ -0,0 +1,337 @@
+        }
+
+    def make_derived(self, **overrides):
+        derived = copy.copy(self)
+        for name, wrapper in overrides.items():
+            base = getattr(derived, name, None)
+            if base is None:
+                raise NameError(f"No method {name!r} exists for type {type(derived).__name__}")
+
+            wrapped = functools.update_wrapper(wrapper(base), base)
+            setattr(derived, name, wrapped)
+
+        derived.mapping = self.mapping.copy()
+        return derived
+
+    def setup_stream_logger(self, stdout, stderr, *, no_warnings=False):
+        stdout_output = NULL_OUTPUT if stdout is None else StreamOutput(stdout, self._allow_color, self._pref_encoding)
+        stderr_output = NULL_OUTPUT if stderr is None else StreamOutput(stderr, self._allow_color, self._pref_encoding)
+
+        self.mapping.update({
+            LogLevel.DEBUG: (
+                stderr_output if self._verbosity is Verbosity.VERBOSE
+                else NULL_OUTPUT),
+            LogLevel.INFO: stdout_output,
+            LogLevel.WARNING: NULL_OUTPUT if no_warnings else stderr_output,
+            LogLevel.ERROR: stderr_output,
+        })
+        return self
+
+    def setup_class_logger(self, logger):
+        debug_logger = ClassOutput(logger.debug)
+        error_logger = ClassOutput(logger.error)
+        warning_logger = ClassOutput(logger.warning)
+
+        self.mapping.update({
+            LogLevel.DEBUG: debug_logger,
+            LogLevel.INFO: debug_logger,
+            LogLevel.WARNING: warning_logger,
+            LogLevel.ERROR: error_logger,
+        })
+        return self
+
+    def setup_logging_logger(self):
+        self.mapping.update({
+            LogLevel.DEBUG: LoggingOutput(logging.DEBUG),
+            LogLevel.INFO: LoggingOutput(logging.INFO),
+            LogLevel.WARNING: LoggingOutput(logging.WARNING),
+            LogLevel.ERROR: LoggingOutput(logging.ERROR),
+        })
+        return self
+
+    def log(self, level, message, *, newline=True, once=False, suppress=False, trace=None, prefix=None):
+        logger = self.mapping.get(level)
+        if not logger or suppress:
+            return
+
+        assert isinstance(message, str)
+
+        if once:
+            if message in self.message_cache:
+                return
+            self.message_cache.add(message)
+
+        if logger.allow_bidi and self._bidi_initalized:
+            message = self._apply_bidi_workaround(message)
+
+        if prefix:
+            message = ' '.join((*map(str, variadic(prefix)), message))
+
+        if level is LogLevel.ERROR and self._verbosity is Verbosity.VERBOSE:
+            message += '\n'
+            if trace is not None:
+                message += str(trace)
+
+            elif sys.exc_info()[0]:  # called from an except block
+                message += traceback.format_exc()
+
+            else:
+                message += ''.join(traceback.format_list(traceback.extract_stack()))
```

Comment:
This is the mentioned compat break for error messages. Previously this was handled as a seperate call, but this just appends to the message.

Do we need to keep compat here? What usecase might break from this?